cou/utils/juju_utils.py

```diff
@@ -442,6 +442,63 @@ async def get_status(self) -> FullStatus:
         model = await self._get_model()
         return await model.get_status()
 
+    async def _use_dispatch(self, unit_name: str) -> bool:
+        """Check if the charm support dispatch the hook or not.
+
+        Legacy and reactive charm allows the operators to directly run hooks
+        inside the charm code directly; while the operator framework uses
+        ./dispatch script to dispatch the hooks. This method check if we could
+        use dispatch to run the hook.
+
+        :param unit_name: Name of the unit to run update-status hook
+        :type unit_name: str
+        :return: true if we can use dispatch
+        :rtype: bool
+        """
+        try:
+            await self.run_on_unit(unit_name, "ls ./dispatch")
+        except CommandRunFailed:
+            return False
+        return True
+
+    async def _use_hooks(self, unit_name: str) -> bool:
+        """Check if the charm support running the hook or not.
+
+        Legacy and reactive charm allows the operators to directly run hooks
+        inside the charm code directly; while the operator framework uses
+        ./dispatch script to dispatch the hooks. This method check if we could
+        run the hook directly.
+
+        :param unit_name: Name of the unit to run update-status hook
+        :type unit_name: str
+        :return: true if we can use hooks
+        :rtype: bool
+        """
+        try:
+            await self.run_on_unit(unit_name, "ls hooks/update-status")
+        except CommandRunFailed:
+            return False
+        return True
+
+    async def update_status(self, unit_name: str) -> None:
+        """Run the update_status hook on the given unit.
+
+        :param unit_name: Name of the unit to run update-status hook
+        :type unit_name: str
+        :raises CommandRunFailed: When update-status hook failed
+        """
+        # For charm written in operator framework
+        if await self._use_dispatch(unit_name):
+            await self.run_on_unit(unit_name, "JUJU_DISPATCH_PATH=hooks/update-status ./dispatch")
+            return
+
+        # For charm written in legacy / reactive framework
+        if await self._use_hooks(unit_name):
+            await self.run_on_unit(unit_name, "hooks/update-status")
+            return
```

Comment:
No, we need separate command to check the file exist or not because update-status hook can result in charm failure. Using above pseudo code we can't distinguish if the command failed because of missing file or it actually failed

Reply:
I don't understand, we can already rely on exit code to know if the file exist or not, this is what the code is doing right now. I thought you mean directly run the hook, that's not working because `run ./dispatch` can be actually failing for update-status, but not failing for missing `dispatch` file.

```
try:
  run ./dispatch
except command-not-found:
  run hooks/update-status
```

Reply:
ah, I think I got you point. 

```
try:
  run ./dispatch
  check stderr if command-not-found raise command-not-found
except command-not-found:
  run hooks/update-status
  check stderr if command-not-found raise command-not-found
except  command-not-found:
  pass

Reply:
updated

Reply:
continue on following comments